src/storage-operators/src/s3_oneshot_sink/parquet.rs

```diff
@@ -7,28 +7,341 @@
-        _connection_details: S3UploadInfo,
-        _sink_id: &GlobalId,
-        _batch: u64,
+        sdk_config: SdkConfig,
+        connection_details: S3UploadInfo,
+        sink_id: &GlobalId,
+        batch: u64,
     ) -> Result<ParquetUploader, anyhow::Error> {
-        anyhow::bail!("Parquet uploading is not yet implemented")
+        match connection_details.format {
+            S3SinkFormat::Parquet => {
+                let builder = ArrowBuilder::new(&connection_details.desc, 1024)?;
+                Ok(ParquetUploader {
+                    desc: connection_details.desc,
+                    sdk_config: Arc::new(sdk_config),
+                    key_manager: S3KeyManager::new(sink_id, &connection_details.uri),
+                    batch,
+                    max_file_size: connection_details.max_file_size,
+                    next_file_index: 0,
+                    builder,
+                    builder_max_size_bytes: std::cmp::min(
+                        connection_details.max_file_size,
+                        DEFAULT_ARROW_BUILDER_BUFFER_SIZE.as_u64(),
+                    ),
+                    active_file: None,
+                })
+            }
+            _ => anyhow::bail!("Expected Parquet format"),
+        }
     }
-    async fn append_row(&mut self, _row: &Row) -> Result<(), anyhow::Error> {
-        anyhow::bail!("Parquet uploading is not yet implemented")
+
+    async fn append_row(&mut self, row: &Row) -> Result<(), anyhow::Error> {
+        self.builder.add_row(row)?;
+
+        if u64::cast_from(self.builder.row_size_bytes()) > self.builder_max_size_bytes {
+            self.flush_builder().await?;
+        }
+        Ok(())
     }
+
     async fn flush(&mut self) -> Result<(), anyhow::Error> {
-        anyhow::bail!("Parquet uploading is not yet implemented")
+        self.flush_builder().await?;
+        if let Some(active) = self.active_file.take() {
+            active.finish().await?;
+        }
+        Ok(())
+    }
+}
+
+impl ParquetUploader {
+    /// Start a new parquet file for upload. Will finish the current file if one is active.
+    async fn start_new_file(&mut self) -> Result<(), anyhow::Error> {
+        if let Some(active_file) = self.active_file.take() {
+            mz_ore::task::spawn(|| "ParquetFile::finish", active_file.finish())
+                .wait_and_assert_finished()
+                .await?;
+        }
+        let object_key = self
+            .key_manager
+            .data_key(self.batch, self.next_file_index, "parquet");
+        self.next_file_index += 1;
+
+        let schema = self.builder.schema();
+        let bucket = self.key_manager.bucket.clone();
+        let sdk_config = Arc::clone(&self.sdk_config);
+        let max_file_size = self.max_file_size;
+        let new_file = mz_ore::task::spawn(
+            || "ParquetFile::new",
+            ParquetFile::new(schema, bucket, object_key, sdk_config, max_file_size),
+        )
+        .wait_and_assert_finished()
```

Comment:
`.run_in_task` from `FutureExt` might be easier!

Reply:
nice - will switch these all to that, much less code!